chain-impl-mockchain/src/ledger/evm.rs

```diff
@@ -1,15 +1,106 @@
 #[cfg(feature = "evm")]
-use chain_evm::state::{AccountTrie, Balance};
+use super::Error;
+#[cfg(feature = "evm")]
+use crate::smartcontract::Contract;
+#[cfg(feature = "evm")]
+use chain_evm::{
+    machine::{Config, Environment, Log, VirtualMachine},
+    state::{AccountTrie, Balance},
+};
 
 #[derive(Default, Clone, PartialEq, Eq)]
 pub struct Ledger {
     #[cfg(feature = "evm")]
     pub(crate) accounts: AccountTrie,
+    #[cfg(feature = "evm")]
+    pub(crate) logs: Vec<Log>,
 }
 
 impl Ledger {
     pub fn new() -> Self {
-        Default::default()
+        Self {
+            #[cfg(feature = "evm")]
+            accounts: Default::default(),
+            #[cfg(feature = "evm")]
+            logs: Default::default(),
+        }
+    }
+    #[cfg(feature = "evm")]
+    pub fn deploy_contract<'runtime>(
+        &mut self,
+        contract: Contract,
+        config: &'runtime Config,
+        environment: &'runtime Environment,
+    ) -> Result<(), Error> {
+        let mut vm = self.virtual_machine(config, environment);
+        match contract {
+            Contract::Create {
+                caller,
+                value,
+                init_code,
+                gas_limit,
+                access_list,
+            } => {
+                //
+                if let Some((new_state, logs)) =
+                    vm.transact_create(caller, value, init_code, gas_limit, access_list, true)
+                {
+                    // update ledger state
+                    self.accounts = new_state.clone();
+                    self.logs.extend_from_slice(logs);
+                }
+                Ok(())
+            }
+            Contract::Create2 {
+                caller,
+                value,
+                init_code,
+                salt,
+                gas_limit,
+                access_list,
+            } => {
+                if let Some((new_state, logs)) = vm.transact_create2(
+                    caller,
+                    value,
+                    init_code,
+                    salt,
+                    gas_limit,
+                    access_list,
+                    true,
+                ) {
+                    // update ledger state
+                    self.accounts = new_state.clone();
+                    self.logs.extend_from_slice(logs);
+                }
+                Ok(())
+            }
+            Contract::Call {
```

Comment:
nit: don't know if this is standard in the evm ecosystem (and forget me if it is) but it sounds a bit off for me that a call to a contract is in a function called `deploy`